File: idli_book/idli_book/page/idli_dashboard/idli_dashboard.py
"""
Idli Dashboard API
Provides data for charts and metrics
"""

import frappe
from frappe import _
from frappe.utils import flt, getdate, add_months, formatdate, nowdate
from datetime import datetime, timedelta

@frappe.whitelist()
def get_dashboard_data(period="This Month"):
	"""Get comprehensive dashboard data"""
	data = {
		"summary_cards": get_summary_cards(period),
		"cash_flow_chart": get_cash_flow_data(period),
		"income_vs_expense": get_income_expense_data(period),
		"top_customers": get_top_customers(period),
		"top_items": get_top_selling_items(period),
		"outstanding_summary": get_outstanding_summary(),
		"bank_balance": get_bank_balance(),
		"recent_transactions": get_recent_transactions()
	}
	return data

def get_summary_cards(period):
	"""Key metrics cards with counts"""
	filters = get_date_filters(period)
	
	# Use single query for revenue with count
	revenue_data = frappe.db.sql("""
		SELECT 
			COALESCE(SUM(grand_total), 0) as total,
			COUNT(*) as count
		FROM `tabIB Sales Invoice`
		WHERE docstatus = 1
		AND invoice_date BETWEEN %s AND %s
	""", (filters['from_date'], filters['to_date']), as_dict=True)[0]
	
	# Use single query for expenses with count
	expenses_data = frappe.db.sql("""
		SELECT 
			COALESCE(SUM(grand_total), 0) as total,
			COUNT(*) as count
		FROM `tabIB Purchase Bill`
		WHERE docstatus = 1
		AND bill_date BETWEEN %s AND %s
	""", (filters['from_date'], filters['to_date']), as_dict=True)[0]
	
	# Cash flow queries
	cash_in = frappe.db.sql("""
		SELECT COALESCE(SUM(amount), 0) as total
		FROM `tabIB Payment`
		WHERE docstatus = 1
		AND payment_type = 'Receive'
		AND payment_date BETWEEN %s AND %s
	""", (filters['from_date'], filters['to_date']), as_dict=True)[0]
	
	cash_out = frappe.db.sql("""
		SELECT COALESCE(SUM(amount), 0) as total
		FROM `tabIB Payment`
		WHERE docstatus = 1
		AND payment_type = 'Pay'
		AND payment_date BETWEEN %s AND %s
	""", (filters['from_date'], filters['to_date']), as_dict=True)[0]
	
	# Calculate profit
	profit = flt(revenue_data.total) - flt(expenses_data.total)
	
	return {
		"revenue": flt(revenue_data.total, 2),
		"invoice_count": revenue_data.count,
		"expenses": flt(expenses_data.total, 2),
		"bill_count": expenses_data.count,
		"profit": flt(profit, 2),
		"cash_in": flt(cash_in.total, 2),
		"cash_out": flt(cash_out.total, 2),
		"net_cash_flow": flt(cash_in.total - cash_out.total, 2)
	}

def get_cash_flow_data(period):
	"""Cash flow chart data (last 12 months)"""
	data = {"labels": [], "cash_in": [], "cash_out": []}
	
	for i in range(11, -1, -1):
		month_date = add_months(nowdate(), -i)
		month_start = getdate(month_date).replace(day=1)
		month_end = add_months(month_start, 1) - timedelta(days=1)
		
		# Cash In
		cash_in = frappe.db.sql("""
			SELECT COALESCE(SUM(amount), 0) as total
			FROM `tabIB Payment`
			WHERE docstatus = 1
			AND payment_type = 'Receive'
			AND payment_date BETWEEN %s AND %s
		""", (month_start, month_end))[0][0]
		
		# Cash Out
		cash_out = frappe.db.sql("""
			SELECT COALESCE(SUM(amount), 0) as total
			FROM `tabIB Payment`
			WHERE docstatus = 1
			AND payment_type = 'Pay'
			AND payment_date BETWEEN %s AND %s
		""", (month_start, month_end))[0][0]
		
		data["labels"].append(month_start.strftime("%b %Y"))
		data["cash_in"].append(flt(cash_in, 2))
		data["cash_out"].append(flt(cash_out, 2))
	
	return data

def get_income_expense_data(period):
	"""Income vs Expense comparison"""
	filters = get_date_filters(period)
	
	# Group by month for the period
	data = {"labels": [], "income": [], "expense": []}
	
	# Get start and end month
	start_date = getdate(filters['from_date'])
	end_date = getdate(filters['to_date'])
	
	current = start_date.replace(day=1)
	while current <= end_date:
		month_end = add_months(current, 1) - timedelta(days=1)
		
		# Income
		income = frappe.db.sql("""
			SELECT COALESCE(SUM(grand_total), 0)
			FROM `tabIB Sales Invoice`
			WHERE docstatus = 1
			AND invoice_date BETWEEN %s AND %s
		""", (current, month_end))[0][0]
		
		# Expense
		expense = frappe.db.sql("""
			SELECT COALESCE(SUM(grand_total), 0)
			FROM `tabIB Purchase Bill`
			WHERE docstatus = 1
			AND bill_date BETWEEN %s AND %s
		""", (current, month_end))[0][0]
		
		data["labels"].append(current.strftime("%b"))
		data["income"].append(flt(income, 2))
		data["expense"].append(flt(expense, 2))
		
		current = add_months(current, 1)
	
	return data

def get_top_customers(period, limit=5):
	"""Top customers by revenue"""
	filters = get_date_filters(period)
	
	customers = frappe.db.sql("""
		SELECT 
			customer,
			SUM(grand_total) as total_revenue,
			COUNT(*) as invoice_count
		FROM `tabIB Sales Invoice`
		WHERE docstatus = 1
		AND invoice_date BETWEEN %s AND %s
		GROUP BY customer
		ORDER BY total_revenue DESC
		LIMIT %s
	""", (filters['from_date'], filters['to_date'], limit), as_dict=True)
	
	return customers

def get_top_selling_items(period, limit=5):
	"""Top selling items"""
	filters = get_date_filters(period)
	
	items = frappe.db.sql("""
		SELECT 
			ii.item_code,
			ii.item_name,
			SUM(ii.quantity) as total_qty,
			SUM(ii.amount) as total_amount
		FROM `tabIB Invoice Item` ii
		INNER JOIN `tabIB Sales Invoice` si ON ii.parent = si.name
		WHERE si.docstatus = 1
		AND si.invoice_date BETWEEN %s AND %s
		GROUP BY ii.item_code
		ORDER BY total_amount DESC
		LIMIT %s
	""", (filters['from_date'], filters['to_date'], limit), as_dict=True)
	
	return items

def get_outstanding_summary():
	"""Outstanding receivables and payables with counts"""
	receivable_data = frappe.db.sql("""
		SELECT 
			COALESCE(SUM(outstanding_amount), 0) as total,
			COUNT(*) as count
		FROM `tabIB Sales Invoice`
		WHERE docstatus = 1
		AND outstanding_amount > 0
	""", as_dict=True)[0]
	
	payable_data = frappe.db.sql("""
		SELECT 
			COALESCE(SUM(outstanding_amount), 0) as total,
			COUNT(*) as count
		FROM `tabIB Purchase Bill`
		WHERE docstatus = 1
		AND outstanding_amount > 0
	""", as_dict=True)[0]
	
	return {
		"receivable": flt(receivable_data.total, 2),
		"receivable_count": receivable_data.count,
		"payable": flt(payable_data.total, 2),
		"payable_count": payable_data.count
	}

def get_bank_balance():
	"""Current bank/cash balance"""
	# Get all Bank and Cash accounts
	bank_balance = frappe.db.sql("""
		SELECT 
			account,
			SUM(debit - credit) as balance
		FROM `tabIB GL Entry`
		WHERE is_cancelled = 0
		AND account IN (
			SELECT account_name 
			FROM `tabIB Chart of Accounts`
			WHERE account_type IN ('Bank', 'Cash')
		)
		GROUP BY account
	""", as_dict=True)
	
	total = sum([flt(b.balance) for b in bank_balance])
	
	return {
		"total": flt(total, 2),
		"accounts": bank_balance
	}

def get_recent_transactions(limit=10):
	"""Recent transactions across all types"""
	transactions = []
	
	# Recent Invoices
	invoices = frappe.get_all("IB Sales Invoice",
		filters={"docstatus": 1},
		fields=["name", "customer", "invoice_date as date", "grand_total as amount", "'Sales Invoice' as type"],
		order_by="creation desc",
		limit=limit//2
	)
	transactions.extend(invoices)
	
	# Recent Payments
	payments = frappe.get_all("IB Payment",
		filters={"docstatus": 1},
		fields=["name", "party as customer", "payment_date as date", "amount", "payment_type as type"],
		order_by="creation desc",
		limit=limit//2
	)
	transactions.extend(payments)
	
	# Sort by date
	transactions = sorted(transactions, key=lambda x: x.date, reverse=True)[:limit]
	
	return transactions

def get_date_filters(period):
	"""Get date range based on period"""
	today = getdate(nowdate())
	
	if period == "Today":
		return {"from_date": today, "to_date": today}
	elif period == "This Week":
		start = today - timedelta(days=today.weekday())
		return {"from_date": start, "to_date": today}
	elif period == "This Month":
		start = today.replace(day=1)
		return {"from_date": start, "to_date": today}
	elif period == "This Quarter":
		quarter_start_month = ((today.month - 1) // 3) * 3 + 1
		start = today.replace(month=quarter_start_month, day=1)
		return {"from_date": start, "to_date": today}
	elif period == "This Year":
		start = today.replace(month=4, day=1)  # Indian FY starts April
		if today.month < 4:
			start = start.replace(year=today.year - 1)
		return {"from_date": start, "to_date": today}
	else:  # Last 30 Days
		start = today - timedelta(days=30)
		return {"from_date": start, "to_date": today}

File: idli_book/idli_book/page/idli_dashboard/idli_dashboard.js
frappe.pages['idli-dashboard'].on_page_load = function (wrapper) {
	new ZohoDashboard(wrapper);
};

class ZohoDashboard {
	constructor(wrapper) {
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Dashboard',
			single_column: true
		});

		this.wrapper = wrapper;
		this.filters = {
			period: 'This Month',
			from_date: frappe.datetime.month_start(),
			to_date: frappe.datetime.get_today()
		};

		this.setup();
	}

	setup() {
		this.add_filters();
		this.render();
		this.load_data();
	}

	add_filters() {
		const me = this;

		// Period Filter
		this.page.add_field({
			label: 'Period',
			fieldtype: 'Select',
			fieldname: 'period',
			options: [
				'Today',
				'This Week',
				'This Month',
				'This Quarter',
				'This Year',
				'Last Month',
				'Last Quarter',
				'Custom'
			],
			default: 'This Month',
			change: function () {
				const val = this.get_value();
				if (val === 'Custom') {
					me.show_custom_dates();
				} else {
					me.set_dates(val);
					me.load_data();
				}
			}
		});

		this.page.set_secondary_action('Refresh', () => this.load_data());
	}

	set_dates(period) {
		const today = frappe.datetime.get_today();
		let from_date, to_date;

		switch (period) {
			case 'Today':
				from_date = to_date = today;
				break;
			case 'This Week':
				from_date = frappe.datetime.week_start();
				to_date = today;
				break;
			case 'This Month':
				from_date = frappe.datetime.month_start();
				to_date = today;
				break;
			case 'This Quarter':
				from_date = frappe.datetime.quarter_start();
				to_date = today;
				break;
			case 'This Year':
				from_date = frappe.datetime.year_start();
				to_date = today;
				break;
			case 'Last Month':
				from_date = frappe.datetime.add_months(frappe.datetime.month_start(), -1);
				to_date = frappe.datetime.month_end(from_date);
				break;
			case 'Last Quarter':
				from_date = frappe.datetime.add_months(frappe.datetime.quarter_start(), -3);
				to_date = frappe.datetime.add_days(frappe.datetime.quarter_start(), -1);
				break;
		}

		this.filters = { period, from_date, to_date };
	}

	show_custom_dates() {
		const me = this;
		const d = new frappe.ui.Dialog({
			title: 'Select Date Range',
			fields: [
				{ fieldtype: 'Date', fieldname: 'from_date', label: 'From', reqd: 1 },
				{ fieldtype: 'Date', fieldname: 'to_date', label: 'To', reqd: 1 }
			],
			primary_action_label: 'Apply',
			primary_action: (v) => {
				me.filters.from_date = v.from_date;
				me.filters.to_date = v.to_date;
				me.load_data();
				d.hide();
			}
		});
		d.show();
	}

	render() {
		const $main = $(this.wrapper).find('.layout-main-section');
		$main.empty().append(`
			<div class="container-fluid px-4 py-3">
				<!-- KPI Cards -->
				<div class="row g-3 mb-4">
					<div class="col-md-3">
						<div class="card border-0 shadow-sm h-100 kpi-card" data-doctype="IB Sales Invoice">
							<div class="card-body">
								<h6 class="text-muted mb-2">Receivables</h6>
								<h3 class="mb-1 kpi-value" id="val-receivable">₹ 0</h3>
								<small class="text-muted">Outstanding invoices</small>
							</div>
						</div>
					</div>
					<div class="col-md-3">
						<div class="card border-0 shadow-sm h-100 kpi-card" data-doctype="IB Purchase Bill">
							<div class="card-body">
								<h6 class="text-muted mb-2">Payables</h6>
								<h3 class="mb-1 kpi-value" id="val-payable">₹ 0</h3>
								<small class="text-muted">Outstanding bills</small>
							</div>
						</div>
					</div>
					<div class="col-md-3">
						<div class="card border-0 shadow-sm h-100 kpi-card">
							<div class="card-body">
								<h6 class="text-muted mb-2">Revenue</h6>
								<h3 class="mb-1 kpi-value" id="val-revenue">₹ 0</h3>
								<small class="text-muted">For selected period</small>
							</div>
						</div>
					</div>
					<div class="col-md-3">
						<div class="card border-0 shadow-sm h-100 kpi-card">
							<div class="card-body">
								<h6 class="text-muted mb-2">Profit</h6>
								<h3 class="mb-1 kpi-value" id="val-profit">₹ 0</h3>
								<small class="text-muted">Net income</small>
							</div>
						</div>
					</div>
				</div>
				
				<!-- Charts Row -->
				<div class="row g-3 mb-4">
					<div class="col-md-8">
						<div class="card border-0 shadow-sm">
							<div class="card-header bg-white border-bottom">
								<h5 class="mb-0">Cash Flow</h5>
							</div>
							<div class="card-body">
								<div id="chart-cashflow"></div>
							</div>
						</div>
					</div>
					<div class="col-md-4">
						<div class="card border-0 shadow-sm">
							<div class="card-header bg-white border-bottom">
								<h5 class="mb-0">Income vs Expense</h5>
							</div>
							<div class="card-body">
								<div id="chart-income"></div>
							</div>
						</div>
					</div>
				</div>
				
				<!-- Tables Row -->
				<div class="row g-3">
					<div class="col-md-6">
						<div class="card border-0 shadow-sm">
							<div class="card-header bg-white border-bottom">
								<h5 class="mb-0">Bank & Cash Accounts</h5>
							</div>
							<div class="card-body p-0">
								<div class="table-responsive">
									<table class="table table-hover mb-0" id="accounts-table">
										<tbody></tbody>
									</table>
								</div>
							</div>
						</div>
					</div>
					<div class="col-md-6">
						<div class="card border-0 shadow-sm">
							<div class="card-header bg-white border-bottom">
								<h5 class="mb-0">Recent Transactions</h5>
							</div>
							<div class="card-body p-0">
								<div class="table-responsive">
									<table class="table table-hover mb-0" id="txn-table">
										<tbody></tbody>
									</table>
								</div>
							</div>
						</div>
					</div>
				</div>
			</div>
			
			<style>
				.kpi-card {
					cursor: pointer;
					transition: transform 0.2s, box-shadow 0.2s;
				}
				.kpi-card:hover {
					transform: translateY(-4px);
					box-shadow: 0 0.5rem 1rem rgba(0,0,0,0.15) !important;
				}
				.kpi-value {
					font-size: 2rem;
					font-weight: 700;
				}
				.card-header h5 {
					font-size: 1rem;
					font-weight: 600;
				}
			</style>
		`);

		// Add click handlers
		const me = this;
		$main.find('.kpi-card').on('click', function () {
			const doctype = $(this).data('doctype');
			if (doctype) {
				frappe.set_route('List', doctype);
			}
		});
	}

	load_data() {
		const me = this;
		frappe.call({
			method: 'idli_book.idli_book.page.idli_dashboard.idli_dashboard.get_dashboard_data',
			args: {
				period: this.filters.period,
				from_date: this.filters.from_date,
				to_date: this.filters.to_date
			},
			freeze: true,
			callback: (r) => {
				if (r.message) me.update(r.message);
			}
		});
	}

	update(data) {
		// Update values
		$('#val-receivable').text(frappe.format(data.outstanding_summary.receivable, { fieldtype: 'Currency' }));
		$('#val-payable').text(frappe.format(data.outstanding_summary.payable, { fieldtype: 'Currency' }));
		$('#val-revenue').text(frappe.format(data.summary_cards.revenue, { fieldtype: 'Currency' }));

		const profit = data.summary_cards.profit;
		$('#val-profit')
			.text(frappe.format(profit, { fieldtype: 'Currency' }))
			.css('color', profit >= 0 ? '#28a745' : '#dc3545');

		// Charts
		if (data.cash_flow_chart && data.cash_flow_chart.labels) {
			new frappe.Chart('#chart-cashflow', {
				data: {
					labels: data.cash_flow_chart.labels,
					datasets: [
						{ name: 'In', values: data.cash_flow_chart.cash_in },
						{ name: 'Out', values: data.cash_flow_chart.cash_out }
					]
				},
				type: 'bar',
				height: 250,
				colors: ['#28a745', '#dc3545']
			});
		}

		if (data.income_vs_expense && data.income_vs_expense.labels) {
			new frappe.Chart('#chart-income', {
				data: {
					labels: data.income_vs_expense.labels.slice(-6),
					datasets: [
						{ name: 'Income', values: data.income_vs_expense.income.slice(-6) },
						{ name: 'Expense', values: data.income_vs_expense.expense.slice(-6) }
					]
				},
				type: 'line',
				height: 250,
				colors: ['#007bff', '#ffc107']
			});
		}

		// Accounts table
		const accounts = data.bank_balance.accounts || [];
		let acc_html = accounts.length === 0
			? '<tr><td class="text-center text-muted p-3">No accounts</td></tr>'
			: accounts.map(a => `
				<tr>
					<td class="p-3">${a.account}</td>
					<td class="p-3 text-end fw-bold" style="color: ${a.balance >= 0 ? '#28a745' : '#dc3545'}">
						${frappe.format(a.balance, { fieldtype: 'Currency' })}
					</td>
				</tr>
			`).join('');
		$('#accounts-table tbody').html(acc_html);

		// Transactions table
		const txns = data.recent_transactions || [];
		let txn_html = txns.length === 0
			? '<tr><td class="text-center text-muted p-3">No transactions</td></tr>'
			: txns.map(t => `
				<tr>
					<td class="p-3">
						<div class="fw-bold">${frappe.datetime.str_to_user(t.date)}</div>
						<small class="text-muted">${t.customer || '-'}</small>
					</td>
					<td class="p-3 text-end">
						<div class="fw-bold">${frappe.format(t.amount, { fieldtype: 'Currency' })}</div>
						<small class="text-muted">${t.type}</small>
					</td>
				</tr>
			`).join('');
		$('#txn-table tbody').html(txn_html);
	}
}
